src/validator/task.rs

```diff
@@ -0,0 +1,942 @@
+    }
+    while height < main_block_height {
+        let block_hash: String = main_client
+            .send_request("getblockhash", &[json!(height)])
+            .map_err(|err| InitialSyncError::Rpc {
+                method: "getblockhash".to_owned(),
+                source: Box::new(err),
+            })?
+            .ok_or(InitialSyncError::GetBlockHash { height })?;
+        let block: String = main_client
+            .send_request("getblock", &[json!(block_hash), json!(0)])
+            .map_err(|err| InitialSyncError::Rpc {
+                method: "getblock".to_owned(),
+                source: Box::new(err),
+            })?
+            .ok_or_else(|| InitialSyncError::GetBlock {
+                block_hash: block_hash.clone(),
+            })?;
+        let block = bitcoin::consensus::encode::deserialize_hex(&block).unwrap();
+        let () = connect_block(&mut rwtxn, dbs, event_tx, &block, height)?;
+        {
+            /*
+            main_client
+                .send_request("invalidateblock", &[json!(block_hash)])
+                .into_diagnostic()?
+                .ok_or(miette!("failed to invalidate block"))?;
+            */
+        }
+        height += 1;
+        let block_hash = {
+            match hex::decode(&block_hash) {
+                Ok(block_hash) => block_hash,
+                Err(err) => {
+                    return Err(InitialSyncError::DecodeBlockHashHex {
+                        block_hash_hex: block_hash,
+                        source: err,
+                    })
+                }
+            }
+            .try_into()
+            .map(BlockHash::from_byte_array)
+            .unwrap()
+        };
+        dbs.current_chain_tip
+            .put(&mut rwtxn, &UnitKey, &block_hash)?;
+    }
+    dbs.current_block_height
+        .put(&mut rwtxn, &UnitKey, &height)?;
+    let () = rwtxn.commit()?;
+    Ok(())
+}
+
+// FIXME: Rewrite all of this to be more readable.
+/// Single iteration of the task loop
+fn task_loop_once(
+    dbs: &Dbs,
+    event_tx: &Sender<Event>,
+    main_client: &RpcClient,
+) -> Result<(), miette::Report> {
+    let mut txn = dbs.write_txn().into_diagnostic()?;
+    let mut height = dbs
+        .current_block_height
+        .try_get(&txn, &UnitKey)
+        .into_diagnostic()?
+        .unwrap_or(0);
+    let main_block_height: u32 = main_client
+        .send_request("getblockcount", &[])
+        .into_diagnostic()?
+        .ok_or(miette!("failed to get block count"))?;
+    if main_block_height == height {
+        return Ok(());
+    }
+    tracing::debug!("Block height: {main_block_height}");
+
+    while height < main_block_height {
+        let block_hash: String = main_client
+            .send_request("getblockhash", &[json!(height)])
+            .into_diagnostic()?
+            .ok_or(miette!("failed to get block hash"))?;
+        let prev_blockhash =
+            bitcoin::consensus::encode::deserialize_hex::<BlockHash>(&block_hash).unwrap();
```

Comment:
This is also incorrect, same byte endianness issue. 

```rust
let prev_blockhash = BlockHash::from_str(&block_hash).unwrap();
```

I guess the tip here was what you're referring to as sus in the comment below?

Reply:
The part I was referring to was 
```
 tracing::debug!("Mainchain tip: {prev_blockhash}");
 ```
 The mainchain tip should not be called `prev_blockhash`.

Reply:
Fixed